test(subscriber): add initial integration tests

hds opened the pull request:
The `console-subscriber` crate has no integration tests. There are some
unit tests, but without very high coverage of features.

Recently, we've found or fixed a few errors which probably could have
been caught by a medium level of integration testing.

However, testing `console-subscriber` isn't straight forward. It is
effectively a tracing subscriber (or layer) on one end, and a gRPC
server on the other end.

This change adds enough of a testing framework to write some initial
integration tests. It is the first step towards closing #450.

Each test comprises 2 parts:
- One or more "expcted tasks"
- A future which will be driven to completion on a dedicated Tokio runtime.

Behind the scenes, a console subscriber layer is created and it's server
part is connected to a duplex stream. The client of the duplex stream
then records incoming updates and reconstructs "actual tasks". The layer
itself is set as the default subscriber for the duration of `block_on`
which is used to drive the provided future to completioin.

The expected tasks have a set of "matches", which is how we find the
actual task that we want to validate against. Currently, the only value
we match on is the task's name.

The expected tasks also have a set of expectations. These are other
fields on the actual task which are validated once a matching task is
found. Currently, the two fields which can have expectations set on them
are the `wakes` and `self_wakes` fields.

So, to construct an expected task, which will match a task with the name
`"my-task"` and then validate that the matched task gets woken once, the
code would be:

```rust
ExpectedTask::default()
    .match_name("my-task")
    .expect_wakes(1);
```

A future which passes this test could be:

```rust
async {
    task::Builder::new()
        .name("my-task")
        .spawn(async {
            tokio::time::sleep(std::time::Duration::ZERO).await
        })
}
```

The full test would then look like:

```rust
fn wakes_once() {
    let expected_task = ExpectedTask::default()
        .match_name("my-task")
        .expect_wakes(1);

    let future = async {
        task::Builder::new()
            .name("my-task")
            .spawn(async {
                tokio::time::sleep(std::time::Duration::ZERO).await
            })
    };

    assert_task(expected_task, future);
}
```

The PR depends on 2 others:
 - #447 which fixes an error in the logic that determines whether a task
   is retained in the aggregator or not.
 - #451 which exposes the server parts and is necessary to allow us to
   connect the instrument server and client via a duplex channel.

This change contains some initial tests for wakes and self wakes which
would have caught the error fixed in #430. Additionally there are tests
for the functionality of the testing framework itself.

hds committed:
test(subscriber): add initial integration tests

The `console-subscriber` crate has no integration tests. There are some
unit tests, but without very high coverage of features.

Recently, we've found or fixed a few errors which probably could have
been caught by a medium level of integration testing.

However, testing `console-subscriber` isn't straight forward. It is
effectively a tracing subscriber (or layer) on one end, and a gRPC
server on the other end.

This change adds enough of a testing framework to write some initial
integration tests. It is the first step towards closing #450.

Each test comprises 2 parts:
- One or more "expcted tasks"
- A future which will be driven to completion on a dedicated Tokio runtime.

Behind the scenes, a console subscriber layer is created and it's server
part is connected to a duplex stream. The client of the duplex stream
then records incoming updates and reconstructs "actual tasks". The layer
itself is set as the default subscriber for the duration of `block_on`
which is used to drive the provided future to completioin.

The expected tasks have a set of "matches", which is how we find the
actual task that we want to validate against. Currently, the only value
we match on is the task's name.

The expected tasks also have a set of expectations. These are other
fields on the actual task which are validated once a matching task is
found. Currently, the two fields which can have expectations set on them
are the `wakes` and `self_wakes` fields.

So, to construct an expected task, which will match a task with the name
`"my-task"` and then validate that the matched task gets woken once, the
code would be:

```rust
ExpectedTask::default()
    .match_name("my-task")
    .expect_wakes(1);
```

A future which passes this test could be:

```rust
async {
    task::Builder::new()
        .name("my-task")
        .spawn(async {
            tokio::time::sleep(std::time::Duration::ZERO).await
        })
}
```

The full test would then look like:

```rust
fn wakes_once() {
    let expected_task = ExpectedTask::default()
        .match_name("my-task")
        .expect_wakes(1);

    let future = async {
        task::Builder::new()
            .name("my-task")
            .spawn(async {
                tokio::time::sleep(std::time::Duration::ZERO).await
            })
    };

    assert_task(expected_task, future);
}
```

The PR depends on 2 others:
 - #447 which fixes an error in the logic that determines whether a task
   is retained in the aggregator or not.
 - #451 which exposes the server parts and is necessary to allow us to
   connect the instrument server and client via a duplex channel.

This change contains some initial tests for wakes and self wakes which
would have caught the error fixed in #430. Additionally there are tests
for the functionality of the testing framework itself.

hds committed:
fixed warning (unneeded `mut`)

hawkw commented on console-subscriber/tests/support/state.rs:
```diff
@@ -0,0 +1,139 @@
+        }
+    }
+
+    /// Block asynchronously until the desired step has been reached.
```
nit: i would probably not say "block asynchronously" --- if this is asynchronous, it doesn't block, by definition. maybe:
```suggestion
    /// Wait asynchronously until the desired step has been reached.
```

hawkw commented on console-subscriber/tests/support/state.rs:
```diff
@@ -0,0 +1,139 @@
+        loop {
+            if self.step >= desired_step {
+                break;
+            }
```
nit: isn't this just a `while` loop?
```suggestion
        while self.step < desired_step {
```

hds replied:
It totally is...

hawkw commented on console-subscriber/tests/support/state.rs:
```diff
@@ -0,0 +1,139 @@
+    }
+
+    /// Check whether the desired step has been reached without blocking.
+    pub(super) fn try_wait_for_step(&mut self, desired_step: TestStep) -> bool {
```
nit: i get why this ended up being named "try_wait_for_step", because it's a version of `wait_for_step` that doesn't wait, but...couldn't we just call it
```suggestion
    /// Returns `true` if the current step is `desired_step` or later.
    pub(super) fn is_step(&mut self, desired_step: TestStep) -> bool {
```
(in particular, I would expect a function called `try_{foo}` to return a `Result`...)

hds replied:
Yep, agree with this.

hawkw commented on console-subscriber/tests/support/state.rs:
```diff
@@ -0,0 +1,139 @@
+    /// # Panics
+    ///
+    /// This method will panic if the test state is not at the step prior to
+    /// `next_step` or if the underlying channel is closed.
```
```suggestion
    /// `next_step`, or if the underlying channel is closed.
```

hawkw commented on console-subscriber/tests/support/state.rs:
```diff
@@ -0,0 +1,139 @@
+            panic!(
+                "cannot advance to previous or current step! current step: {current}, next step: {next_step}",
+                current = self.step);
+        }
```
nit: `if <CONDITION> { panic!(...) }` could be simplified to `assert!(<CONDITION>, ...)`:
```suggestion
        assert!(
            self.step < next_step,
            "cannot advance to previous or current step! current step: {current}, next step: {next_step}",
            current = self.step,
        );
```

hawkw commented on console-subscriber/tests/support/state.rs:
```diff
@@ -0,0 +1,139 @@
+            (TestStep::TestFinished, TestStep::UpdatesRecorded) => {},
+            (_, _) => panic!(
+                "cannot advance more than one step! current step: {current}, next step: {next_step}",
+                current = self.step),
```
nit: slightly weird formatting here, also we could maybe avoid the `current = ` if we just wrote
```suggestion
            (current, _) => panic!("cannot advance more than one step! current step: {current}, next step: {next_step}"),
```

hawkw commented on console-subscriber/tests/support/subscriber.rs:
```diff
@@ -0,0 +1,318 @@
+                        std::io::ErrorKind::Other,
+                        "Client already taken",
+                    ))
+                }
```
nit: isn't this just `Option::ok_or_else`?
```suggestion
                client.ok_or_else(|| std::io::Error::new(std::io::ErrorKind::Other, "Client already taken"))
```

hawkw commented on console-subscriber/tests/support/subscriber.rs:
```diff
@@ -0,0 +1,318 @@
+
+    let mut client_stream = Some(client_stream);
+    let channel = Endpoint::try_from("http://[::]:6669")
+        .expect("Could not create endpoint")
```
we never actually connect to this port in the test, and just use a `DuplexStream` rather than an actual network connection, right?

maybe worth documenting that:
```suggestion
    // Note: we won't actually try to connect to this port on localhost,
    // because we will call `connect_with_connector` with a service that
    // just returns the `DuplexStream`, instead of making an actual 
    // network connection.
    let endpoint = Endpoint::try_from("http://[::]:6669")
        .expect("Could not create endpoint");
    let channel = endpoint
```

hds replied:
Yeah, that's a good point. Thanks!

hawkw commented on console-subscriber/tests/support/subscriber.rs:
```diff
@@ -0,0 +1,318 @@
+            Ok(stream) => break stream.into_inner(),
+            Err(err) => panic!("Client cannot connect to watch updates: {err}"),
+        }
+    };
```
i don't understand why this is a loop? it looks like we will only ever make one `watch_updates` request, since we either break if it's `Ok` or panic if it's `Err`. can't this just be:
```suggestion
    let mut stream = match client.watch_updates(tonic::Request::new(InstrumentRequest {})).await {
        Ok(stream) => stream.into_inner(),
        Err(err) => panic!("Client cannot connect to watch updates: {err}"),
    };
```

hds replied:
Good catch, I think the loop must have been left over from a previous fix attempt that didn't work.

hawkw commented on console-subscriber/tests/support/subscriber.rs:
```diff
@@ -0,0 +1,318 @@
+            Err(e) => {
+                panic!("update stream error: {}", e);
+            }
+        }
```
style nit, take it or leave it: i'm not a huge fan of the amount of "rightward drift" in this code, would something like this work instead?
```suggestion
    while let Some(update) = stream.next().await {
        let update = match update {
            Ok(u) => u,
            Err(e) => panic!("update stream error: {e}"),
        };

        if let Some(task_update) = &update.task_update {
            for new_task in &task_update.new_tasks {
                let id = match new_task.id {
                    Some(id) => id,
                    None => continue,
                };
                let mut actual_task = ActualTask::new(id.id);
                let name = new_task.fields.iter().find_map(|field| {
                    match field.name? {
                        console_api::field::Name::StrName(name) if name == "task.name" => Some(field.value.clone()),
                        _ => None,
                    }
                });
                actual_task.name = match name {
                    Some(Value::DebugVal(value)) => Some(value.clone()),
                    Some(Value::StrVal(value)) => Some(value.clone()),
                    _ => None, // Anything that isn't string-like shouldn't be used as a name.
                };
                tasks.insert(actual_task.id, actual_task);
            }

            for (id, stats) in &task_update.stats_update {
                if let Some(task) = tasks.get_mut(id) {
                    task.wakes = stats.wakes;
                    task.self_wakes = stats.self_wakes;
                }
            }
        }
```

hds replied:
This is much nicer, thank you! I didn't see this comment until after further changes, so I couldn't apply it, but I copied it pretty much as is.

hawkw commented on console-subscriber/tests/support/task.rs:
```diff
@@ -0,0 +1,228 @@
+#[derive(Clone, Debug)]
+pub(crate) struct ExpectedTask {
+    match_name: Option<String>,
+
```
nit: bonus whitespace here:
```suggestion
```

hawkw commented on console-subscriber/tests/support/task.rs:
```diff
@@ -0,0 +1,228 @@
+    /// All matching rules will be run, if they all succeed, then `true` will
+    /// be returned, otherwise `false`.
+    pub(super) fn matches_actual_task(&self, actual_task: &ActualTask) -> bool {
+        if let Some(match_name) = &self.match_name {
```
am I correct that this is saying that if an `ExpectedTask`'s `match_name` is not `Some`, the `ExpectedTask` will basically never do anything of value? why is `match_name` optional, then? couldn't we just always have there be a `match_name`, and have a constructor that takes the name?

hds replied:
My idea is that in the future we will probably (maybe) want to match tasks on things other than the name. For example, I want to match a task associated with a resource or async op. While it's not implemented yet, it seemed to make sense to follow the pattern that a task can match on different things from the outset.

hawkw commented on console-subscriber/tests/support/task.rs:
```diff
@@ -0,0 +1,228 @@
+                    actual: Some(actual_task.clone()),
+                    failure: format!(
+                        "{self}: expected `self_wakes` to be {expected_self_wakes}, but actual was {actual_self_wakes}",
+                        actual_self_wakes = actual_task.self_wakes),
```
weird formatting
```suggestion
                    failure: format!(
                        "{self}: expected `self_wakes` to be {expected_self_wakes}, but actual was {actual_self_wakes}",
                        actual_self_wakes = actual_task.self_wakes,
                    ),
```

hawkw commented on console-subscriber/tests/support/task.rs:
```diff
@@ -0,0 +1,228 @@
+                    actual: Some(actual_task.clone()),
+                    failure: format!(
+                        "{self}: expected `wakes` to be {expected_wakes}, but actual was {actual_wakes}",
+                        actual_wakes = actual_task.wakes),
```
weird formatting:
```suggestion
                    failure: format!(
                        "{self}: expected `wakes` to be {expected_wakes}, but actual was {actual_wakes}",
                        actual_wakes = actual_task.wakes,
                    ),
```

hawkw commented on console-subscriber/tests/support/task.rs:
```diff
@@ -0,0 +1,228 @@
+                expected: self.clone(),
+                actual: Some(actual_task.clone()),
+                failure: format!(
+                    "{self}: no expectations set, if you want to just expect that a matching task is present, use `expect_present()`")
```
weird formatting:
```suggestion
                failure: format!(
                    "{self}: no expectations set, if you want to just expect that a matching task is present, use `expect_present()`",
                ),
```

hawkw commented on console-subscriber/tests/support/task.rs:
```diff
@@ -0,0 +1,228 @@
+            Some(name) => format!("name={name}"),
+            None => "(no fields to match on)".into(),
+        };
+        write!(f, "Task<{fields}>")
```
nit, TIOLI: using angle brackets here feels a little weird to me, since i would expect them to represent a type parameter? maybe:
```suggestion
        write!(f, "Task< {{ {fields} }}")
```

hawkw commented on console-subscriber/tests/support/task.rs:
```diff
@@ -0,0 +1,228 @@
+/// actual tasks on, as well as the expectations that will be used to validate
+/// which the actual task is as expected.
+#[derive(Clone, Debug)]
+pub(crate) struct ExpectedTask {
```
here's a thought: maybe constructing an `ExpectedTask` should capture the `core::panic::Location` where the `ExpectedTask` was constructed? we could have an `ExpectedTask` constructor with `#[track_caller]` on it, and include that location in the failure message. that way, if there are multiple `ExpectedTasks` in a test, it's clear which one was involved in a test failure.

this isn't a blocker for this PR, but it might be nice to add in the future...

hds replied:
That's a really nice idea. I'll look at that in a follow up.

hds replied:
Created an issue #461 to remind me.

hawkw commented on console-subscriber/tests/support/task.rs:
```diff
@@ -0,0 +1,228 @@
+            Some(actual) => write!(
+                f,
+                "Task Validation Failed!\n  Expected Task: {expected:?}\n  Actual Task:   {actual:?}\n  Failure:       {failure}",
+                expected = self.expected, failure = self.failure),
```
nit: weird formatting
```suggestion
                expected = self.expected,
                failure = self.failure,
            ),
```

hawkw commented on console-subscriber/tests/support/task.rs:
```diff
@@ -0,0 +1,228 @@
+            None => write!(
+                f,
+                "Task Validation Failed!\n  Expected Task: {expected:?}\n  Failure:       {failure}",
+                expected = self.expected, failure = self.failure),
```
```suggestion
                "Task Validation Failed!\n  Expected Task: {expected:?}\n  Failure:       {failure}",
                expected = self.expected,
                failure = self.failure,
            ),
```

hawkw reviewed (commented):
overall, this looks great, thanks for working on this! i left a bunch of relatively minor comments, but overall, I'd be happy to merge this change!

hds commented:
> overall, this looks great, thanks for working on this! i left a bunch of relatively minor comments, but overall, I'd be happy to merge this change!

@hawkw Thanks for the review!

Unfortunately, I just realised today that the CI is telling me that I've got a race condition (and I really thought I'd removed all of them) which I'm not seeing locally and is also behaving differently on different platforms - some platforms fail, some platform actually hang forever it seems. So this PR is going to need a bit of work to find and fix that issue.

hds committed:
wait for additional update from subscriber

After the test ends, we were waiting for a single further update before
evaluating the actual tasks (vs. the expected tasks). Now we wait for 2
updates.

hds committed:
single threaded tests to debug

hds committed:
some more debugging stuff

hds committed:
Trace to stdout for CI debugging

sigh.

hds committed:
enable all OSes

hds committed:
bit more tracing

hawkw commented:
@hds is this branch ready for a review now?

hds commented:
@hawkw unfortunately not yet. I've still got a problem with the test run on CI when executing on Ubuntu. It hangs because the message that the test has been run never gets to the instrumentation client.

Work has been busy lately, so getting to the bottom of this problem has been slow going.

hawkw commented:
@hds okay, cool, thanks for letting me know!

hds committed:
output traces from within runtime under test to console

hds committed:
confirming a suspicion

hds committed:
Increase test state channel capacity

hds committed:
maybe avoid a data race

hds committed:
undo all testing changes

This commit has the same changeset as the original commit.

hds committed:
Apply suggestions from code review

Co-authored-by: Eliza Weisman <eliza@buoyant.io>

hds committed:
add signal task at end of test

Rather than relying on all the tasks becoming visible N update
iterations after the test ends, we spawn a signal task which we then
look for. Once the test has completed (which will almost certainly
happen first) and the signal task has been read, we finish parsing the
current update and then finish immediately.

hds committed:
Merge branch 'main' into hds/subscriber-tests

hds committed:
improved rightward drift in update loop

The one in the instrumentation client.

hds commented:
@hawkw This one is ready for re-review when you have a moment. Thanks!

hawkw commented:
> @hawkw This one is ready for re-review when you have a moment. Thanks!

awesome, thanks for all the time you've spent on this!

hawkw commented on console-subscriber/tests/framework.rs:
```diff
@@ -0,0 +1,184 @@
+//! Framework tests
+//!
+//! The tests in this module are here to verify the testing framework itself.
```
including this is a great idea!

hds replied:
I learned the hard way that test frameworks should have tests as well. (-;

hawkw commented on console-subscriber/Cargo.toml:
```diff
@@ -55,6 +55,7 @@ crossbeam-channel = "0.5"
 
 [dev-dependencies]
 tokio = { version = "^1.21", features = ["full", "rt-multi-thread"] }
+tower = "0.4"
```
do we need all the default features of `tower`, or can we disable them and only use the ones we need? this isn't a huge deal, since it's only a dev-only dependency, but it might make building the tests slightly faster. not important.

hds replied:
Good idea, turns out we don't need any features from tower at all!

hawkw commented on console-subscriber/tests/support/mod.rs:
```diff
@@ -0,0 +1,47 @@
+/// This function will panic if the expectations on the expected task are not
+/// met or if a matching task is not recorded.
+#[track_caller]
+#[allow(dead_code)]
```
why do we need the `allow(dead_code)` on these? is this because some test files only use `assert_task` and others only use `assert_tasks`?

hds replied:
Yes, exactly. Because each integration test is compiled individually (or at least, this is common when running a single test), tests that don't use one of these functions caused a warning, which was annoying.

hawkw commented on console-subscriber/tests/support/mod.rs:
```diff
@@ -0,0 +1,47 @@
+/// are not met or if matching tasks are not recorded for all expected tasks.
+#[track_caller]
+#[allow(dead_code)]
+pub(crate) fn assert_tasks<Fut>(expected_tasks: Vec<ExpectedTask>, future: Fut)
+where
+    Fut: Future + Send + 'static,
+    Fut::Output: Send + 'static,
+{
+    run_test(expected_tasks, future)
+}
```
tiny nit, take it or leave it: is there a reason this is a whole additional function, rather than just being a re-export of `run_test`? we could
```rust
pub use subscriber::run_test as assert_tasks;
```
if we want it to be named `assert_tasks` (but also, we could just name the original function that...)

hds replied:
It was more a case of putting the "internal public" functions together to make the documentation clearer. Otherwise we'd have "public" docs here and on the `run_test` function. Not sure what the best practice is in this case to be honest.

hawkw commented on console-subscriber/tests/support/state.rs:
```diff
@@ -0,0 +1,133 @@
+                    // we don't mind being lagged, we'll just get the latest state
+                }
+                Err(RecvError::Closed) => {
+                    panic!("failed to receive current step, waiting for step: {desired_step}, did the test abort?");
```
does this panic indicate an issue in the test framework, or a normal test failure? if this is a test framework issure, can we add something to the panic message that states "this is likely a test framework bug" or similar?

hds replied:
Done.

hawkw commented on console-subscriber/tests/support/state.rs:
```diff
@@ -0,0 +1,133 @@
+
+        self.sender
+            .send(next_step)
+            .expect("failed to send the next test step, did the test abort?");
```
if these panics indicate a bug in the test framework rather than in the test itself, can we add something like:
```suggestion
        assert!(
            self.step < next_step,
            "console-test error: cannot advance to previous or current step! \
            current step: {current}, next step: {next_step}",
            current = self.step,
        );

        match (&self.step, &next_step) {
            (TestStep::Start, TestStep::ServerStarted) |
            (TestStep::ServerStarted, TestStep::ClientConnected) |
            (TestStep::ClientConnected, TestStep::TestFinished) |
            (TestStep::TestFinished, TestStep::UpdatesRecorded) => {},
            (current, _) => panic!(
                "console-test error: test cannot advance more than one step! \
                current step: {current}, next step: {next_step}"
            ),
        }

        self.sender
            .send(next_step)
            .expect(
                "console-test error: failed to send the next test step, \
                did the test abort?"
            );
```

hawkw commented on console-subscriber/tests/support/state.rs:
```diff
@@ -0,0 +1,133 @@
+                    // we don't mind being lagged, we'll just get the latest state
+                }
+                Err(TryRecvError::Closed) => {
+                    panic!("failed to update current step, did the test abort?")
```
is this an internal test framework error, or a test failure?

hds replied:
Internal error, marked it as such.

hawkw commented on console-subscriber/tests/support/subscriber.rs:
```diff
@@ -0,0 +1,323 @@
+    let mut test_state_test = test_state.clone();
+
+    let join_handle = thread::Builder::new()
+        .name("console::subscriber".into())
```
```suggestion
    
    let thread_name = {
        // Include the name of the test thread in the spawned subscriber thread,
        // to make it clearer which test it belongs to.
        let test = thread::current().name().unwrap_or("<unknown test">);
        format!("{test}-console::subscriber")
    }
    let join_handle = thread::Builder::new()
        .name(thread_name)
```

hds replied:
This is a great idea, thanks!

hawkw commented on console-subscriber/tests/support/subscriber.rs:
```diff
@@ -0,0 +1,323 @@
+                actual_tasks
+            })
+        })
+        .expect("console subscriber could not spawn thread");
```
this is an internal test error, right?

hds replied:
Yes, marked as such.

hawkw commented on console-subscriber/tests/support/subscriber.rs:
```diff
@@ -0,0 +1,323 @@
+
+    let join_handle = thread::Builder::new()
+        .name("console::subscriber".into())
+        .spawn(move || {
```
i was going to leave a comment asking why we needed to spawn this stuff in a background thread, but i realized that it's to ensure that any `tracing` spans emitted by the client and console server tasks aren't recorded by the test console subscriber. can we maybe include a comment explaining this? something like:
```suggestion
        // Run the test's console server and client tasks in a separate thread
        // from the main test, ensuring that any `tracing` emitted by the
        // console worker and the client are not collected by the subscriber
        // under test.
        .spawn(move || {
```

hawkw commented on console-subscriber/tests/support/subscriber.rs:
```diff
@@ -0,0 +1,323 @@
+        .expect("client-console error: couldn't spawn aggregator");
+    Server::builder()
+        .add_service(service)
+        .serve_with_incoming(futures::stream::iter(vec![Ok::<_, std::io::Error>(
+            server_stream,
+        )]))
```
very tiny, unimportant nit: could be simplified to just
```suggestion
        .serve_with_incoming(futures::stream::once(Ok::<_, std::io::Error>(server_stream)))
```

hawkw commented on console-subscriber/tests/support/subscriber.rs:
```diff
@@ -0,0 +1,323 @@
+            server_stream,
+        )]))
+        .await
+        .expect("client-console error: couldn't start instrument server.");
```
tiny nit:
```suggestion
        .expect("console-test error: couldn't spawn aggregator");
    Server::builder()
        .add_service(service)
        .serve_with_incoming(futures::stream::iter(vec![Ok::<_, std::io::Error>(
            server_stream,
        )]))
        .await
        .expect("console-test error: couldn't start instrument server.");
```

hawkw commented on console-subscriber/tests/support/subscriber.rs:
```diff
@@ -0,0 +1,323 @@
+            }
+        }))
+        .await
+        .expect("client-console error: couldn't create client");
```
nit: maybe
```suggestion
        .expect("console-test client error: couldn't create client");
```

hawkw commented on console-subscriber/tests/support/subscriber.rs:
```diff
@@ -0,0 +1,323 @@
+            async move {
+                client.ok_or_else(|| {
+                    std::io::Error::new(std::io::ErrorKind::Other, "Client already taken")
+                })
```
nit, take it or leave it: probably better to panic if the client is already taken, since i believe the server will continue accepting connections if it sees a conn error (IIRC?)
```suggestion
                Ok(client.expect("console-test error: client already taken (this shouldn't happen)"))
```

hawkw commented on console-subscriber/tests/support/subscriber.rs:
```diff
@@ -0,0 +1,323 @@
+        .await
+    {
+        Ok(stream) => stream.into_inner(),
+        Err(err) => panic!("Client cannot connect to watch updates: {err}"),
```
is this an internal error or a test failure?

hds replied:
Yes, marked as such.

hawkw commented on console-subscriber/tests/support/subscriber.rs:
```diff
@@ -0,0 +1,323 @@
+    let signal_task = ExpectedTask::default().match_name(END_SIGNAL_TASK_NAME.into());
+    let mut signal_task_read = false;
+    while let Some(update) = stream.next().await {
+        let update = update.expect("update stream error");
```
nit: maybe
```suggestion
        let update = update.expect("console-test error: update stream error");
```

hawkw commented on console-subscriber/tests/support/subscriber.rs:
```diff
@@ -0,0 +1,323 @@
+
+    let mut tasks = HashMap::new();
+
+    let signal_task = ExpectedTask::default().match_name(END_SIGNAL_TASK_NAME.into());
```
this kind of in-band signalling feels a _bit_ sketchy --- would it be better to indicate that the test has ended using a `tokio::sync::oneshot` or something and `select!`ing over either `stream.next()` or the oneshot? that way, we don't have to rely on the console subscriber correctly recording the "end signal task" before we can end the test. 

i'm okay with merging this as-is, if a rewrite would be too time-consuming, since the current approach does seem to _work_. but, i would prefer to separate the end-of-test signal from the actual test tasks...

hawkw replied:
or, if there's a reason we need this to be in-band (is it necessary to ensure that *all* the tasks have definitely been read, rather than just stopping as soon as we trigger a oneshot?) can we add a comment somewhere explaining why this is the best approach?

hds replied:
Yeah, I really didn't want to do this. In fact, the `TestState` struct is supposed to do the job of the oneshot channel you describe. However, when the `TestFinished` state is reached, we don't have all the events from the test yet.

The reason for this is that the update loop in the aggregator tends to yield back to the runtime various times during the first full update. This means that once the test has finished, we have N update loops (even on the client side) until all of the events have been processed, and we don't know the value of N ahead of time (it seems to change from platform to platform too).

https://github.com/tokio-rs/console/blob/d1764046e55a55ce79228c1ce7fdc7153484d601/console-subscriber/src/aggregator/mod.rs#L223

I'll add a comment to explain this in the code.

hawkw commented on console-subscriber/tests/support/subscriber.rs:
```diff
@@ -0,0 +1,323 @@
+                    Some(Value::DebugVal(value)) => Some(value.clone()),
+                    Some(Value::StrVal(value)) => Some(value.clone()),
+                    _ => None, // Anything that isn't string-like shouldn't be used as a name.
+                };
```
i`m normally a big fan of cool iterator tricks, but isn't it simpler to just write something like
```suggestion
                for field in new_task.fields {
                    match field.name.as_ref() {
                        console_api::field::Name::StrName(name) if name == "task.name" => {
                            actual_task.name = match field.value.as_ref() {
                                Value::DebugVal(value) => actual_task.name = Some(value),
                                Value::StrVal(value) => actual_task.name = Some(value),
                                _ => continue;
                            }
                        },
                        _ => {}
                    }
                }
```

hawkw commented on console-subscriber/tests/support/subscriber.rs:
```diff
@@ -0,0 +1,323 @@
+                        console_api::field::Name::StrName(name) if name == "task.name" => {
+                            Some(field.value.clone())
+                        }
+                        _ => None,
```
does this actually work? my assumption was that task names would generally be represented on the wire by metadata indices...i think we need to be handling the case where the name is a metadata index.

hawkw replied:
update: oh, i looked at the way task fields are currently recorded, and it looks like we actually always send the string name...we should probably fix that so that task update messages are more efficient on the wire: https://github.com/tokio-rs/console/blob/d1764046e55a55ce79228c1ce7fdc7153484d601/console-subscriber/src/visitors.rs#L306-L346

but, the current code does work for now, and (i would hope) that changing the way tasks are serialized would break ~all the tests...

hds replied:
Yeah, I wasn't sure about this. This is something that should be improved, and that should be caught by the tests too, as you said.

hds replied:
I created an issue for this improvement: #462.

hawkw commented on console-subscriber/tests/support/subscriber.rs:
```diff
@@ -0,0 +1,323 @@
+        .filter_map(|r| match r {
+            Ok(_) => None,
+            Err(validation_error) => Some(validation_error),
+        })
```
very tiny nit: there's actually a [method on `Result`](https://doc.rust-lang.org/stable/std/result/enum.Result.html#method.err) that just does this:
```suggestion
        .filter_map(Result::err)
```

hds replied:
Nice! Hadn't used that before.

hawkw commented on console-subscriber/tests/support/task.rs:
```diff
@@ -0,0 +1,234 @@
+            ),
+            None => write!(
+                f,
+                "Task Validation Failed!\n  Expected Task: {expected:?}\n  Failure:       {failure}",
```
maybe this message should clearly indicate that no actual task was recorded?

hds replied:
Done.

hawkw reviewed (approved):
overall, this looks good to me! i left a bunch of small suggestions, but none of them are major blockers.

hds committed:
Apply suggestions from code review

Co-authored-by: Eliza Weisman <eliza@buoyant.io>

hds committed:
Apply suggestions from code review (part 2)

Co-authored-by: Eliza Weisman <eliza@buoyant.io>

hds committed:
Update console-subscriber/tests/support/subscriber.rs

Co-authored-by: Eliza Weisman <eliza@buoyant.io>

hds committed:
Suggestions from code review

Thanks!

hds commented:
> overall, this looks good to me! i left a bunch of small suggestions, but none of them are major blockers.

@hawkw Thank you so much for all these suggestions! I really appreciate the effort, the PR is much better for them.
